src/functions.php

```diff
@@ -561,5 +563,7 @@ function select_server_for_aggregate_write_stage(Manager $manager, array &$optio
         throw $serverSelectionError;
     }
 
+    assert($server instanceof Server);
```

Comment:
Necessary as psalm trips over our fancy logic above. We can assume that if the original `select_server` call didn't return a server, `$serverSelectionError` would be set, so the `return` wouldn't be reached. This logic seems to be too advanced for psalm, so using this to silence the error.

Reply:
Works for me. Since this differs from some other instances where you needed to add `assert()` in that psalm fails to detect a potential assignment via `catch`, I wonder if they might be interested in a bug report to address this.

Reply:
I also checked whether this is expected behaviour based on a simplified example: https://psalm.dev/r/b1102ead61

Reply:
Nice that they have a web-based tool available. Is that example suitable to open a bug report? I assume this isn't desired behavior.

Beyond the scope of this PR to be sure, so no need to follow up here.

Reply:
https://github.com/vimeo/psalm/issues/8361 looks related and I added the snippet there. We'll see what comes of it.